libs/remote_storage/src/azure_blob.rs

```diff
@@ -0,0 +1,379 @@
+//! Azure Blob Storage wrapper
+
+use std::num::NonZeroU32;
+use std::{borrow::Cow, collections::HashMap, io::Cursor};
+
+use super::REMOTE_STORAGE_PREFIX_SEPARATOR;
+use anyhow::Result;
+use azure_core::request_options::{MaxResults, Metadata, Range};
+use azure_core::Header;
+use azure_storage::StorageCredentials;
+use azure_storage_blobs::prelude::ClientBuilder;
+use azure_storage_blobs::{
+    blob::operations::GetBlobBuilder,
+    prelude::{BlobClient, ContainerClient},
+};
+use futures_util::StreamExt;
+use http_types::StatusCode;
+use tokio::io::AsyncRead;
+use tracing::debug;
+
+use crate::s3_bucket::RequestKind;
+use crate::{
+    AzureConfig, ConcurrencyLimiter, Download, DownloadError, RemotePath, RemoteStorage,
+    StorageMetadata,
+};
+
+pub struct AzureBlobStorage {
+    client: ContainerClient,
+    prefix_in_container: Option<String>,
+    max_keys_per_list_response: Option<NonZeroU32>,
+    concurrency_limiter: ConcurrencyLimiter,
+}
+
+impl AzureBlobStorage {
+    pub fn new(azure_config: &AzureConfig) -> Result<Self> {
+        debug!(
+            "Creating azure remote storage for azure container {}",
+            azure_config.container_name
+        );
+
+        let account =
+            std::env::var("AZURE_STORAGE_ACCOUNT").expect("missing AZURE_STORAGE_ACCOUNT");
+        let access_key =
+            std::env::var("AZURE_STORAGE_ACCESS_KEY").expect("missing AZURE_STORAGE_ACCESS_KEY");
```

Comment:
Keys should be optional: we should let the Azure SDK use environment for auth (equivalent of what we do with S3 and IAM roles) in production

Reply:
I couldn't find a "use environment" equivalent for the Azure SDK crates [`azure_core`](https://docs.rs/azure_core/0.16.0/azure_core/), [`azure_storage`](https://docs.rs/azure_storage/latest/azure_storage/) and [`azure_storage_blobs`](https://docs.rs/azure_storage_blobs/0.16.0/azure_storage_blobs/) . The [example for azure_storage_blobs](https://docs.rs/azure_storage_blobs/0.16.0/azure_storage_blobs/) also uses env vars.

Ultimately this "use environment" functionality also just reads a bunch of env vars.

Reply:
Azure tends to call this whole business "Active Directory" (which is confusing, as it's only tangentially related to the traditional windows active directory).

The flow we want is probably something like https://learn.microsoft.com/en-gb/azure/active-directory/develop/v2-oauth2-auth-code-flow via https://crates.io/crates/azure_identity , then pass it into the ABS library https://docs.rs/azure_storage/0.16.0/azure_storage/struct.StorageCredentials.html#method.bearer_token

It doesn't have to be in this PR, but let's make sure we are creating tickets for the work between this PR and being prod-ready.

Reply:
I see, makes sense. Will file an issue as follow-up.

Reply:
this is very similar to aws, just in the sdk it hides it by using EnvironmentVariableCredentialsProvider